wp-includes/sqlite-ast/class-wp-sqlite-driver.php

```diff
@@ -0,0 +1,2601 @@
+	 */
+	public function get_last_mysql_query(): ?string {
+		return $this->last_mysql_query;
+	}
+
+	/**
+	 * Get SQLite queries executed for the last MySQL query.
+	 *
+	 * @return array{ sql: string, params: array }[]
+	 */
+	public function get_last_sqlite_queries(): array {
+		return $this->last_sqlite_queries;
+	}
+
+	/**
+	 * Get the auto-increment value generated for the last query.
+	 *
+	 * @return int|string
+	 */
+	public function get_insert_id() {
+		$last_insert_id = $this->pdo->lastInsertId();
+		if ( is_numeric( $last_insert_id ) ) {
+			$last_insert_id = (int) $last_insert_id;
+		}
+		return $last_insert_id;
+	}
+
+	/**
+	 * Translate and execute a MySQL query in SQLite.
+	 *
+	 * A single MySQL query can be translated into zero or more SQLite queries.
+	 *
+	 * @param string $query              Full SQL statement string.
+	 * @param int    $fetch_mode         PDO fetch mode. Default is PDO::FETCH_OBJ.
+	 * @param array  ...$fetch_mode_args Additional fetch mode arguments.
+	 *
+	 * @return mixed Return value, depending on the query type.
+	 *
+	 * @throws WP_SQLite_Driver_Exception When the query execution fails.
+	 */
+	public function query( string $query, $fetch_mode = PDO::FETCH_OBJ, ...$fetch_mode_args ) {
+		$this->flush();
+		$this->pdo_fetch_mode   = $fetch_mode;
+		$this->last_mysql_query = $query;
+
+		try {
+			// Parse the MySQL query.
+			$lexer  = new WP_MySQL_Lexer( $query );
+			$tokens = $lexer->remaining_tokens();
+
+			$parser = new WP_MySQL_Parser( self::$mysql_grammar, $tokens );
+			$ast    = $parser->parse();
+
+			if ( null === $ast ) {
+				throw $this->new_driver_exception( 'Failed to parse the MySQL query.' );
+			}
+
+			// Handle transaction commands.
+
+			/*
+			 * [GRAMMAR]
+			 * beginWork: BEGIN_SYMBOL WORK_SYMBOL?
+			 */
+			$child = $ast->get_first_child();
+			if ( $child instanceof WP_Parser_Node && 'beginWork' === $child->rule_name ) {
+				$this->begin_transaction();
+				return true;
+			}
+
+			if ( $child instanceof WP_Parser_Node && 'simpleStatement' === $child->rule_name ) {
+				/*
+				 * [GRAMMAR]
+				 * transactionOrLockingStatement:
+				 *   transactionStatement | savepointStatement | lockStatement | xaStatement
+				 */
+				$subchild = $child->get_first_child_node( 'transactionOrLockingStatement' );
+				if ( null !== $subchild ) {
+					$tokens = $subchild->get_descendant_tokens();
+					$token1 = $tokens[0];
+					$token2 = $tokens[1] ?? null;
+					if (
+						WP_MySQL_Lexer::START_SYMBOL === $token1->id
+						&& WP_MySQL_Lexer::TRANSACTION_SYMBOL === $token2->id
+					) {
+						$this->begin_transaction();
+						return true;
+					}
+
+					if (
+						WP_MySQL_Lexer::BEGIN_SYMBOL === $token1->id
+					) {
+						$this->begin_transaction();
+						return true;
+					}
+
+					if (
+						WP_MySQL_Lexer::COMMIT_SYMBOL === $token1->id
+					) {
+						$this->commit();
+						return true;
+					}
+
+					if (
+						WP_MySQL_Lexer::ROLLBACK_SYMBOL === $token1->id
```

Comment:
Should we bale out if we see `ROLLBACK TO SAVEPOINT` until we can support that?

Reply:
Actually we do support them internally so maybe baling out would be as easy as preserving the savepoint semantics

Reply:
Ticket: https://github.com/Automattic/sqlite-database-integration/issues/14